protocol/testing/e2e/gov/add_new_market_test.go

```diff
@@ -0,0 +1,256 @@
+	perptypes "github.com/dydxprotocol/v4-chain/protocol/x/perpetuals/types"
+	pricestypes "github.com/dydxprotocol/v4-chain/protocol/x/prices/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAddNewMarketProposal(t *testing.T) {
+	testMarketParam1001 := pricestest.GenerateMarketParamPrice(
+		pricestest.WithId(1001),
+	)
+	testClobPair1001 := clobtest.GenerateClobPair(
+		clobtest.WithId(1001),
+		clobtest.WithPerpetualId(1001),
+		clobtest.WithStatus(clobtypes.ClobPair_STATUS_INITIALIZING),
+	)
+	testPerpetual1001 := perptest.GeneratePerpetual(
+		perptest.WithId(1001),
+		perptest.WithMarketId(1001),
+	)
+	msgUpdateClobPair1001ToActive := &clobtypes.MsgUpdateClobPair{
+		Authority: authtypes.NewModuleAddress(delaymsgtypes.ModuleName).String(),
+		ClobPair: *clobtest.GenerateClobPair(
+			clobtest.WithId(1001),
+			clobtest.WithPerpetualId(1001),
+			clobtest.WithStatus(clobtypes.ClobPair_STATUS_ACTIVE),
+		),
+	}
+
+	tests := map[string]struct {
+		proposedMsgs             []sdk.Msg
+		updateClobDelayBlocks    uint32
+		expectSubmitProposalFail bool
+		expectedProposalStatus   govtypesv1.ProposalStatus
+	}{
+		"Success with 4 standard messages": {
+			proposedMsgs: []sdk.Msg{
+				&pricestypes.MsgCreateOracleMarket{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Params:    testMarketParam1001.Param,
+				},
+				&perptypes.MsgCreatePerpetual{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Params:    testPerpetual1001.Params,
+				},
+				&clobtypes.MsgCreateClobPair{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					ClobPair:  *testClobPair1001,
+				},
+				&delaymsgtypes.MsgDelayMessage{
+					Authority:   authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Msg:         encoding.EncodeMessageToAny(t, msgUpdateClobPair1001ToActive),
+					DelayBlocks: 10,
+				},
+			},
+			updateClobDelayBlocks:  10,
+			expectedProposalStatus: govtypesv1.ProposalStatus_PROPOSAL_STATUS_PASSED,
+		},
+		"Proposals execution fails due to incorrectly ordered messages": {
+			proposedMsgs: []sdk.Msg{
+				&pricestypes.MsgCreateOracleMarket{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Params:    testMarketParam1001.Param,
+				},
+				// Create clob pair before creating perpetual, which will fail.
+				&clobtypes.MsgCreateClobPair{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					ClobPair:  *testClobPair1001,
+				},
+				&perptypes.MsgCreatePerpetual{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Params:    testPerpetual1001.Params,
+				},
+				&delaymsgtypes.MsgDelayMessage{
+					Authority:   authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Msg:         encoding.EncodeMessageToAny(t, msgUpdateClobPair1001ToActive),
+					DelayBlocks: 10,
+				},
+			},
+			updateClobDelayBlocks:  10,
```

Comment:
nit: can we just use the delayblock value in the above `MsgDelayMessage`? why do we need to explicitly define this?

Reply:
We can, but then we are making assumption on the index of the `MsgDelayMessage` in the array. So IMO it's cleaner and more explicit to define this